### ISSUE TITLE
Option to use original column for default locale

### ISSUE THREAD

**Issue by wvengen** [shioyama/mobility#335]
Proposal: add an option to use the original column for the default locale, instead of putting the default locale in backend storage.

## Context
When adding Mobility to an existing application (with no other model translation infrastructure in place), there are usually existing columns that need translations. As far as I could see, Mobility expects all translations to live in backend storage, meaning that the original column is abandoned (or perhaps used once in a migration, after which is it abandoned - #315).

When there are many references to the original column, or if there are other applications using the same database (like SQL queries in an external tool), they all need to be rewritten. This can be a large barrier to adoption, and make it harder to change the used Mobility backend later on.

## Suggested solution
Add an option to allow the use of the original column for the default locale.

**Comment by shioyama** [comment]
You should be able to achieve this with:

```ruby
class Post < Application Record
  def title(**opts)
    if I18n.locale == I18n.default_locale
      super(**opts, super: true)
    else
      super
    end  
end
```

I haven't tried the code above, but basically the idea is that `super` is a Mobility option which allows you to call through to whatever the original method was (defined by ActiveRecord for the `title` column in this case). So when you call `post.title(super: true)` you get the actual value of the `title` column. But you only want to do that when the locale is the default locale, hence the conditional.

The code above may require some tweaking.

**Comment by shioyama** [comment]
I agree though that this is a common request and that there should eventually be an option or plugin for it.

**Comment by wvengen** [comment]
Thanks for your quick response! And for the code example. One could indeed write a concern that adapts translated attributes like that (and the setter, and perhaps some others).
Great to hear this suggestion is welcome.

**Comment by shioyama** [comment]
Glad that helped!

Just added the setter code. Also realized that this would not work with querying, and that one would be trickier to get working.

Anyway I'll leave this open for now as a feature, although I doubt I'll be getting to it for a while.

**Comment by yourtallness** [comment]
Ran into a similar issue when I added mobility to my app today.

All product names became empty because there are no translation records yet and the gem does not use the original column.

For existing apps are we supposed to create a data migration to create translation records for the source language?

**Comment by shioyama** [comment]
> For existing apps are we supposed to create a data migration to create translation records for the source language?

Yes, there is basically no support for doing this at the moment, although it's something I've thought about now and again. It should eventually be supported, but it's just never gotten the priority to make it happen. I don't really see that changing any time soon unless somebody from the community wants to take it on.

**Comment by mattboldt** [comment]
@shioyama Bumping this thread, as I have a similar idea to retain existing model structure while having Mobility tables act as sort of "alternate translations" v.s. storing the "primary" language in a new row.

After some tinkering, I was able to build a backend that's more-or-less the `KeyValue` backend with a sort of `Column` backend as a fallback. I'm able to read, write, and query on my translated column (simply named `name`). If the current locale is the default, it reverts to `Column` backend behavior. Else, we call `super` to the `KeyValue` backend.

I'm very new to Mobility and Localization in general, so I'm not sure how airtight this solution is. I'm curious if you have recommendations on how best to test, or if an obvious issue jumps out at you.

```ruby
require 'mobility/backends/active_record/key_value'

module Mobility
  module Backends
    class HybridKeyValue < ActiveRecord::KeyValue
      def read(locale, **)
        if locale.to_s == I18n.default_locale
          model.read_attribute(attribute)
        else
          super
        end
      end

      def write(locale, value, **)
        if locale.to_s == I18n.default_locale
          model.write_attribute(attribute, value)
        else
          super
        end
      end

      def self.build_node(attr, locale)
        if locale.to_s == I18n.default_locale
          # is the MobilityExpressions plugin necessary here?
          model_class.arel_table[attr].extend(Plugins::Arel::MobilityExpressions)
        else
          super
        end
      end
    end

    register_backend(:hybrid_key_value, HybridKeyValue)
  end
end

# translated model:
class Program < ActiveRecord::Base
  extend Mobility
  translates :name, type: :string
end

# schema:
create_table "programs", id: :integer, options: "..." do |t|
  t.string "name"
end
```
**Rails**: 5.2.4.3
**Mobility**: 1.1.1

My config:
```ruby
Mobility.configure do
  plugins do
    backend :hybrid_key_value
    active_record
    reader
    writer
    backend_reader
    query
    cache
    dirty
    presence
    fallthrough_accessors
    locale_accessors [:en, :fr]
  end
end
```

**Comment by shioyama** [comment]
Oh very nice! Super cool that you used `build_node` to also support querying.

> # is the MobilityExpressions plugin necessary here?

I'm actually not sure, I'd have to refresh my memory.

**Comment by shioyama** [comment]
Actually you know I think this could be a plugin instead of a backend. Basically a plugin which returns the original column instead of the Mobility backend if the locale is the default locale. That would make it much more powerful.

**Comment by mattboldt** [comment]
I'm struggling to figure out how to turn this into a plugin, or how that would improve things. 

> plugin which returns the original column instead of the Mobility backend

I can't figure out where I'd be doing this. Would I be overriding a method like this? https://github.com/shioyama/mobility/blob/7d512ee771991fb107fbdd164ef957f5dfd4984a/lib/mobility/plugins/active_record/query.rb#L45-L52

And, if the backend itself is skipped entirely, would that affect other plugins like `backend_reader` and `locale_accessors`? If the goal here is to maintain duck typing, do you think it would be disruptive to sometimes return native AR objects/methods, but Mobility backends for other languages?

Sorry if this is a lot to unpack. I'm afraid I'm running out of time to tinker with this, and I need to come up with a solid solution to "migrate" a large application to have localizable models. My thinking was that using the existing columns as a fallback would be the least disruptive for existing data, but I don't want to shoot myself in the foot here if the better-tested solution is just to use plain `KeyValue` and [build migration tasks for each translatable model](https://github.com/ilyakatz/data-migrate).

**Comment by mattboldt** [comment]
Ok so here's where I'm at w/ the backend approach. Seems to work well.

```ruby
module Events
  class Program < ActiveRecord::Base
    extend Mobility
    translates :name, type: :string
    default_scope { i18n }
  end
end

Mobility.with_locale(:en) { Events::Program.last.name }
  # Events::Program Load (0.4ms)  SELECT  `events_programs`.* FROM `events_programs` ORDER BY `events_programs`.`id` DESC LIMIT 1
=> "English Event Name"

Mobility.with_locale(:fr) { Events::Program.last.name }
  # Events::Program Load (0.4ms)  SELECT  `events_programs`.* FROM `events_programs` ORDER BY `events_programs`.`id` DESC LIMIT 1
  # Mobility::Backends::ActiveRecord::KeyValue::StringTranslation Load (0.3ms)  SELECT `mobility_string_translations`.* FROM `mobility_string_translations` WHERE `mobility_string_translations`.`translatable_id` = 1 AND `mobility_string_translations`.`translatable_type` = 'Events::Program' AND `mobility_string_translations`.`key` = 'name'
=> "French Event Name"

Mobility.with_locale(:en) { Events::Program.where(name: 'English Event Name') }
  # Events::Program Load (0.4ms)  SELECT `events_programs`.* FROM `events_programs` WHERE `events_programs`.`name` = 'English Event Name'
=> [#<Events::Program ...>]

Mobility.with_locale(:fr) { Events::Program.where(name: 'French Event Name') }
  # Events::Program Load (0.5ms)  SELECT `events_programs`.* FROM `events_programs` INNER JOIN `mobility_string_translations` `Events::Program_name_fr_string_translations` ON `Events::Program_name_fr_string_translations`.`key` = 'name' AND `Events::Program_name_fr_string_translations`.`locale` = 'fr' AND `Events::Program_name_fr_string_translations`.`translatable_type` = 'Events::Program' AND `Events::Program_name_fr_string_translations`.`translatable_id` = `events_programs`.`id` WHERE `Events::Program_name_fr_string_translations`.`value` = 'French Event Name'
=> [#<Events::Program ...>]

Mobility.with_locale(:en) { pr = Events::Program.last; pr.name; pr.name_backend.send(:cache)  }
# Events::Program Load (0.4ms)  SELECT  `events_programs`.* FROM `events_programs` ORDER BY `events_programs`.`id` DESC LIMIT 1
=> {}

Mobility.with_locale(:fr) { pr = Events::Program.last; pr.name; pr.name_backend.send(:cache)  }
  # Events::Program Load (0.4ms)  SELECT  `events_programs`.* FROM `events_programs` ORDER BY `events_programs`.`id` DESC LIMIT 1
  # Mobility::Backends::ActiveRecord::KeyValue::StringTranslation Load (0.4ms)  SELECT `mobility_string_translations`.* FROM `mobility_string_translations` WHERE `mobility_string_translations`.`translatable_id` = 1 AND `mobility_string_translations`.`translatable_type` = 'Events::Program' AND `mobility_string_translations`.`key` = 'name'
=> {:fr=> #<Mobility::Backends::ActiveRecord::KeyValue::StringTranslation... }
```
Calling `update` on the default language behaves a little weird. Looks like it's trying to load the translation, none are found, and then it falls back to the normal update:
```ruby
Mobility.with_locale(:en) { Events::Program.last.update(name: 'English Event Name Updated') }
# Events::Program Load (0.3ms)  SELECT  `events_programs`.* FROM `events_programs` ORDER BY `events_programs`.`id` DESC LIMIT 1
   # (0.3ms)  BEGIN
  # Mobility::Backends::ActiveRecord::KeyValue::StringTranslation Load (1.2ms)  SELECT `mobility_string_translations`.* FROM `mobility_string_translations` WHERE `mobility_string_translations`.`translatable_id` = 1 AND `mobility_string_translations`.`translatable_type` = 'Events::Program' AND `mobility_string_translations`.`key` = 'name'
# Events::Program Update (2.2ms)  UPDATE `events_programs` SET `updated_at` = '2021-04-06 00:03:28', `name` = 'English Event Name Updated' WHERE `events_programs`.`id` = 1
   # (1.4ms)  COMMIT
```

-----

Here's my backend in its entirety:

```ruby
require 'mobility/backends/active_record/key_value'

module Mobility
  # -
  module Backends
    #
    # Implements the {Mobility::Backends::KeyValue} backend with a Column fallback for default locales.
    #
    class HybridKeyValue < ActiveRecord::KeyValue
      def read(locale, **)
        if self.class.use_column_fallback?(locale)
          model.read_attribute(attribute)
        else
          super
        end
      end

      def write(locale, value, **)
        if self.class.use_column_fallback?(locale)
          model.write_attribute(attribute, value)
        else
          super
        end
      end

      class << self
        def build_node(attr, locale)
          if use_column_fallback?(locale)
            model_class.arel_table[attr]
          else
            super
          end
        end

        def apply_scope(relation, predicate, locale = Mobility.locale, invert: false)
          if use_column_fallback?(locale)
            relation
          else
            super
          end
        end

        def use_column_fallback?(locale)
          locale.to_s == 'en' || locale.to_s == 'en_us'
        end
      end
    end

    register_backend(:hybrid_key_value, HybridKeyValue)
  end
end
```

**Comment by shioyama** [comment]
> I'm struggling to figure out how to turn this into a plugin, or how that would improve things.

No worries, that can also be done later! Looks like great work.

One thing to consider is what happens if you have a uniqueness validator on the attribute. Mobility has a plugin for this which generally works well, but it will change how the normal column-based uniqueness validation works, might make it less efficient or introduce edge cases.

**Comment by shioyama** [comment]
> Calling update on the default language behaves a little weird. Looks like it's trying to load the translation, none are found, and then it falls back to the normal update:

Not sure, but that might be because of this:

https://github.com/shioyama/mobility/blob/21755ee3b86d3418c1d77ba4180aa6a5ff47ebca/lib/mobility/backends/active_record/key_value.rb#L173-L177

**Comment by mattboldt** [comment]
Seriously, thank you for all your (very quick) input! It's very much appreciated. I look forward to expanding on this idea and contributing back to the gem. For now I'm gonna implement the backend I have and see how things go.

**Comment by shioyama** [comment]
Hi folks, I quickly took what's here and made a plugin out of it in #512. It has no tests and probably needs a bit more work on the details, but the benefit of a plugin is that it should work for _any_ backend, not just KeyValue.

I don't really have time to workout detials for a while so if anyone is interested please give it a try and comment or contribute (tests would be especially great :slightly_smiling_face: ).

**Comment by mugar** [comment]
Hi @shioyama , just run into the same issue of adding mobility to existing application. i will try your plungin and let you know

**Comment by shioyama** [comment]
#512 hsa been merged and this has been released in v1.2.0 as the new `column_fallback` plugin.
